openprescribing/api/views_measures.py

```diff
@@ -149,144 +162,141 @@ def measure_numerators_by_org(request, format=None):
 
 @api_view(['GET'])
 def measure_by_ccg(request, format=None):
-    measure = request.query_params.get('measure', None)
+    measure_id = request.query_params.get('measure', None)
     orgs = utils.param_to_list(request.query_params.get('org', []))
     tags = [x for x in request.query_params.get('tags', '').split(',') if x]
-    params = []
-    query = 'SELECT mv.month AS date, mv.numerator, mv.denominator, '
-    query += 'mv.calc_value, mv.percentile, mv.cost_savings, '
-    query += 'mv.pct_id, pc.name as pct_name, measure_id, '
-    query += 'ms.name, ms.title, ms.description, '
-    query += 'ms.why_it_matters, ms.denominator_short, '
-    query += 'ms.numerator_short, '
-    query += 'ms.url, ms.is_cost_based, ms.is_percentage, '
-    query += 'ms.low_is_good '
-    query += "FROM frontend_measurevalue mv "
-    query += "JOIN frontend_pct pc ON "
-    query += "(mv.pct_id=pc.code AND pc.org_type = 'CCG') "
-    query += "JOIN frontend_measure ms ON mv.measure_id=ms.id "
-    query += "WHERE pc.close_date IS NULL AND "
-    if orgs:
-        query += "("
-        for i, org in enumerate(orgs):
-            query += "mv.pct_id=%s "
-            params.append(org)
-            if (i != len(orgs) - 1):
-                query += ' OR '
-        query += ") AND "
-    query += 'mv.practice_id IS NULL '
-    if measure:
-        query += "AND mv.measure_id=%s "
-        params.append(measure)
-    for tag in tags:
-        query += "AND %s = ANY(ms.tags) "
-        params.append(tag)
-    query += "ORDER BY mv.pct_id, measure_id, date"
-    data = utils.execute_query(query, [params])
+
+    org_Q = Q()
+    for org in orgs:
+        org_Q |= Q(pct_id=org)
+
+    qs = MeasureValue.objects.\
+        select_related('pct', 'measure').\
+        filter(
+            org_Q,
+            pct__org_type='CCG',
+            pct__close_date__isnull=True,
+            practice_id__isnull=True,
+        ).\
+        order_by('pct_id', 'measure_id', 'month')
+
+    if measure_id:
+        qs = qs.filter(measure_id=measure_id)
+
+    if tags:
+        valid_tags = ['core', 'lowpriority']
```

Comment:
Could probably not bother for now, but given (a) it's repeated and (b) we have a live plan to add more tags, it would be nicer to either have this in settings, or check against values in `Measure.objects.values_list('tags').distinct()` or whatever it would be.

I'm worried we won't pick this up in tests when we add new tags and everything will break.

Reply:
Good point.  I think we need better testing here full stop.